Add a triaging guide for suspected rocMLIR failures. (#3227)
ROCm/AMDMIGraphX: docs/dev/triage-rocmlir.rst

Issue Triaging Guide for suspected rocMLIR issue
================================================

This document serves as a guide to narrow down whether a bug is due to
rocMLIR backend of MIGraphX.

There are broadly 3 categories of bugs that can be due to rocMLIR.

> 1. [B1]rocMLIR compilation bug

Reviewer: Another debugging check could reverting rocMLIR SHA. 
If there was a known working SHA before try building MIGraphX against that and see if that works without any problems. 
and then try newer rocMLIR SHA and see if it fails or passes with MIGraphX. That could indicate if problem lies in rocMLIR or MIGraphX

Reviewer 2: Yeah

But also, conversely, holding the rocMLIR SHA constant at the newest one while debugging can be useful

Author: changing rocMLIR SHA feels like something rocMLIR team should be doing...
I m not sure whether it is advisable to try older SHAs where bugs are fixed in the later ones.

Therefore, I think disabling rocMLIR might be enough with the latest SHA in most cases to see whether its rocMLIR vs MIGraphX bug. (Yes, there can be complex cases where rocMLIR offloads creating a different graph structure exposing a MIGraphX bug but with current infrastructure there is no good way to handle that -- we can have a separate discussion how to improve debuggability) 

I d go further and say unless there is a warning from the following, we should not be changing rocMLIR SHA in debugging.  https://github.com/ROCm/AMDMIGraphX/blob/3e65032266c013e840fd08a2f41623a822e1c538/src/targets/gpu/mlir.cpp#L42-L43

2. [B2]rocMLIR runtime failure
3. [B3]accuracy issue from rocMLIR

Step 1 - Use MIGRAPHX_DISABLE_MLIR=1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> First see if the bug persist disabling MLIR. If so, it is highly likely

Reviewer 3: ```suggestion
First see if the bug persists after disabling MLIR. If so, it is highly likely
```

> First see if the bug persist disabling MLIR. If so, it is highgly likely

Reviewer: ```suggestion
First see if the bug persist disabling MLIR. If so, it is highly likely
```

it is not a MLIR bug but rather a MIGraphX bug.

If you dont see a failure, please proceed.

Step 2 - See if its a B1 - rocMLIR compilation bug
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> So MLIR has two pipelines that is used to generate a kernel: - highlevel
> pipeline - backend pipeline

Reviewer 3: ```suggestion
MLIR has two pipelines that are used to generate a kernel: highlevel
pipeline and backend pipeline
```

like as follows :

::

   module {
     func.func @mlir_convolution(%arg0: !migraphx.shaped<2x8x3x3xf32, 72x9x3x1>, %arg1: !migraphx.shaped<1x8x4x4xf32, 128x16x4x1>) -> !migraphx.shaped<1x2x2x2xf32, 8x4x2x1> attributes {arch = "gfx90a:sramecc+:xnack-", enable_splitk_for_tuning = true, kernel = "mixr", num_cu = 110 : i64} {
       %0 = migraphx.convolution %arg1, %arg0 {dilation = [1, 1], group = 1 : i64, padding = [0, 0, 0, 0], padding_mode = 0 : i64, stride = [1, 1]} : <1x8x4x4xf32, 128x16x4x1>, <2x8x3x3xf32, 72x9x3x1> -> <1x2x2x2xf32, 8x4x2x1>
       return %0 : !migraphx.shaped<1x2x2x2xf32, 8x4x2x1>
     }
   }

Please provide only the *SINGLE* module that fails and not a dump of
them for rocMLIR team figure out which is failing.

Step 2.2 If the backend pipeline fails for *all solutions*, you should see error that starts with
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``No solutions provided for mlir_*`` or
``No valid tuned compilation for mlir_*``

NOTE 1 : Currently MIGraphX does not run applicability pipeline of
rocMLIR. Therefore, you will see them for inapplicable tuning params as
well.

> NOTE 2 : Ignore ``MLIR backend compilation failed: "`` if dont see

Reviewer: ```suggestion
NOTE 2 : Ignore ``MLIR backend compilation failed: "`` if you dont see
```

in with ``MIGRAPHX_TRACE_MLIR=2``

For e.g. :

::

   # This is MIGRAPHX module

   arg2 = @param:arg2 -> float_type, {1, 5, 3}, {15, 3, 1}
   arg1 = @param:arg1 -> float_type, {1, 4, 3}, {12, 3, 1}
   arg0 = @param:arg0 -> float_type, {1, 5, 4}, {20, 4, 1}
   @3 = dot(arg0,arg1) -> float_type, {1, 5, 3}, {15, 3, 1}
   @4 = add(@3,arg2) -> float_type, {1, 5, 3}, {15, 3, 1}
   @5 = @return(@4)

   # This is the MLIR module

   module {
     func.func @mlir_dot_add(%arg0: !migraphx.shaped<1x5x4xf32, 20x4x1>, %arg1: !migraphx.shaped<1x4x3xf32, 12x3x1>, %arg2: !migraphx.shaped<1x5x3xf32, 15x3x1>) -> !migraphx.shaped<1x5x3xf32, 15x3x1> attributes {arch = "gfx90a:sramecc+:xnack-", enable_splitk_for_tuning = true, kernel = "mixr", num_cu = 110 : i64} {
       %0 = migraphx.dot %arg0, %arg1 : <1x5x4xf32, 20x4x1>, <1x4x3xf32, 12x3x1> -> <1x5x3xf32, 15x3x1>
       %1 = migraphx.add %0, %arg2 : <1x5x3xf32, 15x3x1>, <1x5x3xf32, 15x3x1> -> <1x5x3xf32, 15x3x1>
       return %1 : !migraphx.shaped<1x5x3xf32, 15x3x1>
     }
   }

Then individually create MIGraphX program that only has the MIGRAPHX
module Then indiviually ``driver verify`` them to see which is the

> failing module.

Reviewer 2: I don't think MIGraphX is able to do this easily

Author: yes but someone has to do it, right ?

Reviewer: https://github.com/ROCm/AMDMIGraphX/pull/3182/files# 

With this PR it would be possible to dump MIGraphX programs for each module as MXR and then migraphx-driver can driver verify them.
